File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValidBSTRecursive(root, None, None)
    
    def isValidBSTRecursive(self, root: Optional[TreeNode], min_val, max_val) -> bool:
        if root == None:
            return True
        
        if min_val != None and root.val <= min_val:
                return False
        if max_val != None and max_val <= root.val:
                return False
        
        left_valid =  self.isValidBSTRecursive(root.left, min_val, root.val)
        right_valid = self.isValidBSTRecursive(root.right, root.val, max_val)
        
        return left_valid and right_valid
        
        

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/*
func isValidBST(root *TreeNode) bool {
    return isValidBSTRecursive(root, nil, nil)
}

func isValidBSTRecursive(root *TreeNode, min, max *int) bool {
    if root == nil {
        return true
    }
    
    if min != nil && root.Val <= *min {
        return false
    }
    
    if max != nil && *max <= root.Val {
        return false
    }
    
    leftValid := isValidBSTRecursive(root.Left, min, &root.Val)
    rightValid := isValidBSTRecursive(root.Right, &root.Val, max)
    
    return leftValid && rightValid
}
*/

func isValidBST(root *TreeNode) bool {
    visitedOrder := inorderTraversal(root)
    
    for i := 1; i < len(visitedOrder); i++ {
        if visitedOrder[i] <= visitedOrder[i-1] {
            return false
        }
    }
    
    return true
}

func inorderTraversal(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    
    leftVisited := inorderTraversal(root.Left)
    visited := append(leftVisited, root.Val)
    rightVisited := inorderTraversal(root.Right)
    
    return append(visited, rightVisited...)
}